### PR TITLE
DEPENDENT: API: Expose bitcoinconsensus_verify_header() in libconsensus

### PR DESCRIPTION
**Pull request by jtimon**
To extend libconsensus' functionality, I propose to add `bitcoinconsensus_verify_header()` the partial checks `CheckBlockHeader()` and `ContextualCheckBlockHeader` could be exposed as well.

Unlike VerifyScript, VerifyBlockHeader depends on the nodes storage, as it requires data from the blockchain previous to the header being verified. How to exactly interface that data with VerifyBlockHeader can be a delicate matter and it should be of course open to discussion. Here there's just one proposal.

The offending dependency is CBlockIndex in chain.o, specially CBlockIndex::pprev which is a pointer to another object. All the other dependencies unacceptable for libconsensus have been removed first, to make things easier for someone who wants to propose to solve that final problem in some other way.

Dependencies:
-MOVEONLY: Move most consensus function declarations to consensus/consensus.h #6048
-Consensus: Refactor: Consensus version of CheckBlockHeader() #6035
-Separate CValidationState from main #5669
-Chainparams: Refactor: Decouple IsSuperMajority from Params() #5968
-Consensus: Decouple ContextualCheckBlockHeader from checkpoints #5975
-Consensus: Turn CBlockIndex::GetMedianTimePast into independent function #6009
-Consensus: Consensus version of pow functions #6037

### CODE REVIEW

**Comment by jtimon** [comment]
Needed rebase

File: src/consensus/blockverify.cpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+#include "consensus/consensus.h"
+
+#include "arith_uint256.h"
+#include "consensus/validation.h"
+#include "primitives/block.h"
+#include "tinyformat.h"
+
+#include <algorithm>  
+
+static const unsigned int MEDIAN_TIME_SPAN = 11;
```
Comment (Diapolo) [inline src/consensus/blockverify.cpp]: IMHO this could get a comment.

File: src/consensus/validation.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        chRejectCode = chRejectCodeIn;
+        strRejectReason = strRejectReasonIn;
+        corruptionPossible = corruptionIn;
+        if (mode == MODE_ERROR)
+            return ret;
+        nDoS += level;
+        mode = MODE_INVALID;
+        return ret;
+    }
+    bool Invalid(bool ret = false,
```
Comment (Diapolo) [inline src/consensus/validation.h]: This looks strange/ugly, you could IMHO let clang-format cleanup this pull :)?

File: src/main.cpp [hunk 3]
```diff
@@ -1206,6 +1200,21 @@ bool IsInitialBlockDownload()
     return state;
 }
 
+static arith_uint256 GetBlockProof(const CBlockIndex& block)
+{
+    arith_uint256 bnTarget;
+    bool fNegative;
+    bool fOverflow;
+    bnTarget.SetCompact(block.nBits, &fNegative, &fOverflow);
```
Comment (Diapolo) [inline src/main.cpp]: Block can never be NULL here?

File: src/miner.cpp [hunk 4]
```diff
@@ -78,17 +80,33 @@ class TxPriorityCompare
     }
 };
 
+uint32_t GetNextWorkRequiredLog(const CBlockIndex* pindexLast, const CBlockHeader *pblock, const Consensus::Params& consensusParams)
+{
+    uint32_t nextChallenge = GetNextWorkRequired(pindexLast, pblock, consensusParams, GetPrevIndex);
+    /// debug print
+    LogPrintf("GetNextWorkRequired RETARGET\n");
```
Comment (Diapolo) [inline src/miner.cpp]: Function name changed, now ``GetNextWorkRequiredLog``, perhaps use __func__?

File: src/consensus/validation.h (hunk 2 above)
Comment (jtimon) [inline src/consensus/validation.h]: This is from a moveonly commit in #5669 
Hopefully we will do clang-format project wise for 0.11. @laanwj what were the requirements for that to happen?

File: src/consensus/blockverify.cpp (hunk 1 above)
Comment (jtimon) [inline src/consensus/blockverify.cpp]: Feel free to suggest something. Probably better to comment in the smaller PRs for now as this branch will likely require more rebases and force pushes. In this case, the relevant PR is #6009

File: src/main.cpp (hunk 3 above)
Comment (jtimon) [inline src/main.cpp]: This is a move-only commit as well.

File: src/miner.cpp (hunk 4 above)
Comment (jtimon) [inline src/miner.cpp]: I don't undesrtand what you mean. There's two functions with this, one for miner and one for everyone else. I'm happy to remove the miner version if it's fine to remove the logging there.
Anyway, I will create a smaller PR only adapting the pow functions to consensus.

**Comment by jtimon** [comment]
I realized this is mostly unreadable until something similar to #6048 is merged.
I will close this for now but I will still maintain the list of dependencies.

**Pull request closed without merging**